File: ios/src_library/React Native/RCTContextMenuButton/RCTContextMenuButtonManager.m
//
//  RCTContextMenuButtonManager.m
//  IosContextMenuExample
//
//  Created by Dominic Go on 10/28/20.
//  Copyright © 2020 Facebook. All rights reserved.
//

#import <React/RCTViewManager.h>
 

@interface RCT_EXTERN_MODULE(RCTContextMenuButtonManager, RCTViewManager)

// -----------------------------------
// MARK: Props - RN Comp Props: Events
// -----------------------------------

// TODO: WIP

RCT_EXPORT_VIEW_PROPERTY(onMenuWillShow  , RCTBubblingEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onMenuWillHide  , RCTBubblingEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onMenuWillCancel, RCTBubblingEventBlock);

RCT_EXPORT_VIEW_PROPERTY(onMenuDidShow  , RCTBubblingEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onMenuDidHide  , RCTBubblingEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onMenuDidCancel, RCTBubblingEventBlock);

RCT_EXPORT_VIEW_PROPERTY(onPressMenuItem   , RCTBubblingEventBlock);

// -----------------------------------
// MARK: Props - RN Comp Props: Values
// -----------------------------------

RCT_EXPORT_VIEW_PROPERTY(menuConfig, NSDictionary);

@end

File: ios/src_library/React Native/RCTContextMenuButton/RCTContextMenuButtonManager.swift
//
//  RCTContextMenuButtonManager.swift
//  IosContextMenuExample
//
//  Created by Dominic Go on 10/28/20.
//  Copyright © 2020 Facebook. All rights reserved.
//

import UIKit;


@objc(RCTContextMenuButtonManager)
class RCTContextMenuButtonManager: RCTViewManager {
  
  override static func requiresMainQueueSetup() -> Bool {
    return true;
  };
  
  override func view() -> UIView! {
    if #available(iOS 14, *) {
      return RCTContextMenuButton(bridge: self.bridge);
      
    } else if #available(iOS 13, *) {
      return RCTContextMenuView(bridge: self.bridge);
      
    } else {
      return RCTView();
    };
  };
  
  @objc func dismissMenu(_ node: NSNumber) {
    DispatchQueue.main.async {
      guard let view =  self.bridge.uiManager.view(forReactTag: node)
      else { return };
      
      if #available(iOS 14, *),
         let contextMenuButton = view as? RCTContextMenuButton {

        contextMenuButton.dissmissMenu();
        
      } else if #available(iOS 13, *),
                let contextMenuView = view as? RCTContextMenuView {
        
        contextMenuView.dissmissMenu();
      };
    };
  };
};
